File: RunLogKotlin/app/src/main/java/com/runlog/data/AppConfig.java
package com.runlog.data;

public class AppConfig {
    public static final String DEFAULT_SCHOOL_HOST = "http://sports.aiyyd.com:8080";
    public static final String DEFAULT_API_HOST = "http://sports.aiyyd.com:9001/api";
    public static final String DEFAULT_APP_EDITION = "3.6.2";
    public static final String DEFAULT_PUBLIC_KEY = "BDdKFsuBf51UObke1pEgfER17biBg/5r8slqE4s8oOa8lVesWgIUxsRc+AmZ72GcuJ56f7avnyJe3CJY4n00LU4=";
    public static final String DEFAULT_CIPHER_KEY = "JXhWGZjmhhXN+nt8nLpNxA==";
    public static final String DEFAULT_CIPHER_KEY_ENCRYPTED = "BGfbsG9EkXz5KeCva8E0MisBeS6bhBEDId3VXeIuBoiBMZU0Mosv7PqKsvqxZ3PjkUlsjzh09Se629SWW45XP4TIUeXoLpYzgk5fAMbg0VNVnXuLH9xVzdHAeM+1qJrgvwwkwio85/DnrP1aArvVQrw3N4xd5tugqQ==";
    public static final String DEFAULT_MD5_KEY = "pie0hDSfMRINRXc7s1UIXfkE";

    public String schoolId = "";
    public String schoolName = "";
    public String schoolHost = DEFAULT_SCHOOL_HOST;
    public String appEdition = DEFAULT_APP_EDITION;
    public String platform = "android";
    public String publicKey = DEFAULT_PUBLIC_KEY;
    public String cipherKey = DEFAULT_CIPHER_KEY;
    public String cipherKeyEncrypted = DEFAULT_CIPHER_KEY_ENCRYPTED;
    public String md5Key = DEFAULT_MD5_KEY;
    public String token = "";
    public String deviceId = "";
    public String deviceName = "";
    public String uuid = "";
    public String username = "";
    public String password = "";
    public String userType = "1";
    public String runMode = "DATA";
    public String pickMode = "RANDOM";
    public boolean driftEnabled = false;
    public String selectedTask = "";
    public boolean draft = true;

    public AppConfig copy() {
        AppConfig out = new AppConfig();
        out.schoolId = schoolId;
        out.schoolName = schoolName;
        out.schoolHost = schoolHost;
        out.appEdition = appEdition;
        out.platform = platform;
        out.publicKey = publicKey;
        out.cipherKey = cipherKey;
        out.cipherKeyEncrypted = cipherKeyEncrypted;
        out.md5Key = md5Key;
        out.token = token;
        out.deviceId = deviceId;
        out.deviceName = deviceName;
        out.uuid = uuid;
        out.username = username;
        out.password = password;
        out.userType = userType;
        out.runMode = runMode;
        out.pickMode = pickMode;
        out.driftEnabled = driftEnabled;
        out.selectedTask = selectedTask;
        out.draft = draft;
        return out;
    }

    public void merge(CandidateConfig candidate) {
        if (!blank(candidate.schoolId)) schoolId = candidate.schoolId;
        if (!blank(candidate.schoolName)) schoolName = candidate.schoolName;
        if (!blank(candidate.schoolHost)) schoolHost = candidate.schoolHost;
        if (!blank(candidate.appEdition)) appEdition = candidate.appEdition;
        if (!blank(candidate.platform)) platform = candidate.platform;
        if (!blank(candidate.token)) token = candidate.token;
        if (!blank(candidate.deviceId)) deviceId = candidate.deviceId;
        if (!blank(candidate.deviceName)) deviceName = candidate.deviceName;
        if (!blank(candidate.uuid)) uuid = candidate.uuid;
    }

    static boolean blank(String value) {
        return value == null || value.trim().isEmpty();
    }
}

File: RunLogKotlin/core/src/main/kotlin/com/runlog/core/AppConfig.kt
package com.runlog.core

data class AppConfig(
    val schoolId: String = "",
    val schoolName: String = "",
    val schoolHost: String = DefaultSchoolHost,
    val appEdition: String = DefaultAppEdition,
    val platform: String = "android",
    val publicKey: String = "",
    val cipherKey: String = "",
    val md5Key: String = "",
    val token: String = "",
    val deviceId: String = "",
    val deviceName: String = "",
    val uuid: String = ""
) {
    fun merged(candidate: CandidateConfig): AppConfig = copy(
        schoolHost = candidate.schoolHost.ifBlank { schoolHost },
        appEdition = candidate.appEdition.ifBlank { appEdition },
        platform = candidate.platform.ifBlank { platform },
        token = candidate.token.ifBlank { token },
        deviceId = candidate.deviceId.ifBlank { deviceId },
        deviceName = candidate.deviceName.ifBlank { deviceName },
        uuid = candidate.uuid.ifBlank { uuid }
    )
}

const val DefaultSchoolHost = "http://sports.aiyyd.com:8080"
const val DefaultApiHost = "http://sports.aiyyd.com:9001/api"
const val DefaultAppEdition = "3.6.2"

enum class RunMode { QUICK, TABLE }
enum class PickMode { RANDOM, SPECIFIED }

data class CandidateConfig(
    val schoolHost: String = "",
    val appEdition: String = "",
    val platform: String = "",
    val token: String = "",
    val deviceId: String = "",
    val deviceName: String = "",
    val uuid: String = "",
    val requestHost: String = "",
    val requestLine: String = ""
)

data class ConfigValidationResult(
    val valid: Boolean,
    val missingFields: List<String>,
    val warnings: List<String>
) {
    val summary: String
        get() = buildString {
            append(if (valid) "基础配置通过" else "基础配置未通过")
            if (missingFields.isNotEmpty()) append("\n缺失: ${missingFields.joinToString()}")
            if (warnings.isNotEmpty()) append("\n提示: ${warnings.joinToString()}")
        }
}

object ConfigValidator {
    fun validate(config: AppConfig): ConfigValidationResult {
        val missing = mutableListOf<String>()
        fun require(name: String, value: String, minLength: Int = 1) {
            if (value.trim().length < minLength) missing += name
        }

        require("school_host", config.schoolHost)
        require("platform", config.platform)
        require("app_edition", config.appEdition)
        require("token", config.token, 10)
        require("device_id", config.deviceId, 12)
        require("device_name", config.deviceName)
        require("uuid", config.uuid, 8)

        val warnings = mutableListOf<String>()
        if (!config.platform.equals("android", ignoreCase = true)) {
            warnings += "platform 建议为 android"
        }
        if (config.appEdition != DefaultAppEdition) {
            warnings += "当前版本 ${config.appEdition}，建议核对是否仍为 $DefaultAppEdition"
        }
        if (config.schoolHost.contains(":9001")) {
            warnings += "跑步接口不应使用 9001/api，已知可用值为 $DefaultSchoolHost"
        }
        if (!looksLikeUuid(config.uuid)) {
            warnings += "uuid 格式不像标准 UUID，但仍允许继续检测接口"
        }

        return ConfigValidationResult(missing.isEmpty(), missing, warnings)
    }

    fun validateCandidate(candidate: CandidateConfig): ConfigValidationResult =
        validate(
            AppConfig(
                schoolHost = candidate.schoolHost,
                appEdition = candidate.appEdition,
                platform = candidate.platform,
                token = candidate.token,
                deviceId = candidate.deviceId,
                deviceName = candidate.deviceName,
                uuid = candidate.uuid
            )
        )

    private fun looksLikeUuid(value: String): Boolean {
        val text = value.trim()
        return text.length >= 32 && text.count { it == '-' } in 0..4
    }
}
